### <a name="change-in-behavior-for-taskwaitall-methods-with-time-out-arguments"></a><span data-ttu-id="4184a-101">Изменения в поведении для Task.WaitAll методы с аргументами времени ожидания</span><span class="sxs-lookup"><span data-stu-id="4184a-101">Change in behavior for Task.WaitAll methods with time-out arguments</span></span>

|   |   |
|---|---|
|<span data-ttu-id="4184a-102">Подробные сведения</span><span class="sxs-lookup"><span data-stu-id="4184a-102">Details</span></span>|<span data-ttu-id="4184a-103">Поведение Task.WaitAll в был выполнен более согласованной .NET 4.5.In .NET Framework 4, эти методы поведение было непоследовательным.</span><span class="sxs-lookup"><span data-stu-id="4184a-103">Task.WaitAll behavior was made more consistent in .NET 4.5.In the .NET Framework 4, these methods behaved inconsistently.</span></span> <span data-ttu-id="4184a-104">По истечении времени ожидания, если одна или несколько задач были завершены или отменены до вызова метода, метод вызывал исключение <xref:System.AggregateException?displayProperty=name>.</span><span class="sxs-lookup"><span data-stu-id="4184a-104">When the time-out expired, if one or more tasks were completed or canceled before the method call, the method threw an <xref:System.AggregateException?displayProperty=name> exception.</span></span> <span data-ttu-id="4184a-105">По истечении времени ожидания, если ни одна задача не была завершена или отменена до вызова метода, однако одна или несколько задач входили в эти состояния после вызова метода, метод возвращал значение false.</span><span class="sxs-lookup"><span data-stu-id="4184a-105">When the time-out expired, if no tasks were completed or canceled before the method call, but one or more tasks entered these states after the method call, the method returned false.</span></span><br/><br/><span data-ttu-id="4184a-106">В .NET Framework 4.5, эти перегрузки метода теперь возвращают значение false, если все задачи по-прежнему выполняются после истечения интервала времени ожидания, и вызывают исключение <xref:System.AggregateException?displayProperty=name> только в том случае, если входная задача была отменена (независимо от того, была ли она до или после метода вызов), а не выполняются никакие другие задачи.</span><span class="sxs-lookup"><span data-stu-id="4184a-106">In the .NET Framework 4.5, these method overloads now return false if any tasks are still running when the time-out interval expired, and they throw an <xref:System.AggregateException?displayProperty=name> exception only if an input task was cancelled (regardless of whether it was before or after the method call) and no other tasks are still running.</span></span>|
|<span data-ttu-id="4184a-107">Предложение</span><span class="sxs-lookup"><span data-stu-id="4184a-107">Suggestion</span></span>|<span data-ttu-id="4184a-108">Если <xref:System.AggregateException?displayProperty=name> перехвачено с точки зрения обнаружение задачу, которая была отменена до вызова метода Метод WaitAll вызываемого кода вместо этого делать же обнаружения через свойство IsCanceled (например:. Any(t =&gt; t.IsCanceled)) с момента .NET 4.6 только вызывает исключение в этом случае, если выполняются все ожидающей задачи до времени ожидания.</span><span class="sxs-lookup"><span data-stu-id="4184a-108">If an <xref:System.AggregateException?displayProperty=name> was being caught as a means of detecting a task that was cancelled prior to the WaitAll call being invoked, that code should instead do the same detection via the IsCanceled property (for example: .Any(t =&gt; t.IsCanceled)) since .NET 4.6 will only throw in that case if all awaited tasks are completed prior to the timeout.</span></span>|
|<span data-ttu-id="4184a-109">Область</span><span class="sxs-lookup"><span data-stu-id="4184a-109">Scope</span></span>|<span data-ttu-id="4184a-110">Дополнительный номер</span><span class="sxs-lookup"><span data-stu-id="4184a-110">Minor</span></span>|
|<span data-ttu-id="4184a-111">Версия</span><span class="sxs-lookup"><span data-stu-id="4184a-111">Version</span></span>|<span data-ttu-id="4184a-112">4.5</span><span class="sxs-lookup"><span data-stu-id="4184a-112">4.5</span></span>|
|<span data-ttu-id="4184a-113">Тип</span><span class="sxs-lookup"><span data-stu-id="4184a-113">Type</span></span>|<span data-ttu-id="4184a-114">Среда выполнения</span><span class="sxs-lookup"><span data-stu-id="4184a-114">Runtime</span></span>|
|<span data-ttu-id="4184a-115">Затронутые API</span><span class="sxs-lookup"><span data-stu-id="4184a-115">Affected APIs</span></span>|<ul><li><xref:System.Threading.Tasks.Task.WaitAll(System.Threading.Tasks.Task[],System.Int32)?displayProperty=nameWithType></li><li><xref:System.Threading.Tasks.Task.WaitAll(System.Threading.Tasks.Task[],System.Int32,System.Threading.CancellationToken)?displayProperty=nameWithType></li><li><xref:System.Threading.Tasks.Task.WaitAll(System.Threading.Tasks.Task[],System.TimeSpan)?displayProperty=nameWithType></li></ul>|

